chia/full_node/subscriptions.py

```diff
@@ -1,144 +1,179 @@
 from __future__ import annotations
 
 import logging
-from dataclasses import dataclass, field
-from typing import Dict, List, Set
+import sqlite3
+from typing import List, Set
 
 from chia.types.blockchain_format.sized_bytes import bytes32
 
 log = logging.getLogger(__name__)
 
 
-# The PeerSubscriptions class is essentially a multi-index container. It can be
-# indexed by peer_id, coin_id and puzzle_hash.
-@dataclass(frozen=True)
 class PeerSubscriptions:
-    # TODO: use NewType all over to describe these various uses of the same types
-    # Puzzle Hash : Set[Peer ID]
-    _coin_subscriptions: Dict[bytes32, Set[bytes32]] = field(default_factory=dict, init=False)
-    # Puzzle Hash : Set[Peer ID]
-    _ph_subscriptions: Dict[bytes32, Set[bytes32]] = field(default_factory=dict, init=False)
-    # Peer ID: Set[Coin ids]
-    _peer_coin_ids: Dict[bytes32, Set[bytes32]] = field(default_factory=dict, init=False)
-    # Peer ID: Set[puzzle_hash]
-    _peer_puzzle_hash: Dict[bytes32, Set[bytes32]] = field(default_factory=dict, init=False)
-    # Peer ID: subscription count
-    _peer_sub_counter: Dict[bytes32, int] = field(default_factory=dict, init=False)
-
-    def has_ph_subscription(self, ph: bytes32) -> bool:
-        return ph in self._ph_subscriptions
-
-    def has_coin_subscription(self, coin_id: bytes32) -> bool:
-        return coin_id in self._coin_subscriptions
-
-    def add_ph_subscriptions(self, peer_id: bytes32, phs: List[bytes32], max_items: int) -> Set[bytes32]:
-        """
-        returns the puzzle hashes that were actually subscribed to. These may be
-        fewer than requested in case:
-        * there are duplicate puzzle_hashes
-        * some puzzle hashes are already subscribed to
-        * the max_items limit is exceeded
-        """
+    _db_conn: sqlite3.Connection
+
+    def __init__(self) -> None:
+        self._db_conn = sqlite3.connect(":memory:")
+
+        with self._db_conn as conn:
+            log.info("DB: Creating peer subscription tables")
+            conn.execute(
+                """
+                CREATE TABLE puzzle_subscriptions (
+                    id INT PRIMARY KEY,
+                    peer_id BLOB,
+                    puzzle_hash BLOB,
+                    UNIQUE (peer_id, puzzle_hash)
+                )
+                """
+            )
+            conn.execute(
+                """
+                CREATE TABLE coin_subscriptions (
+                    id INT PRIMARY KEY,
+                    peer_id BLOB,
+                    coin_id BLOB,
+                    UNIQUE (peer_id, coin_id)
+                )
+                """
```

Comment:
These tables look like they may actually store duplicates of coin-IDs and puzzle-hashes. Are you confident this is more memory efficient?

I would also expect that you would need an index for both `peer_id`, `coin_id` and `puzzle_hash` for both of these tables.